File: python/backtracking/combination_sum.py
# https://leetcode.com/problems/combination-sum/
class Solution:
    def combinationSum(self, candidates: list[int], target: int) -> list[list[int]]:
        
        def backtrack(start, builder: list, curr_sum):
            if curr_sum > target:
                return 
            if curr_sum == target:
                combinations.append(builder.copy())
                return 

            for i in range(start, len(candidates)):
                builder.append(candidates[i])
                backtrack(i, builder, curr_sum + candidates[i])
                builder.pop()

        combinations = []
        candidates.sort()
        backtrack(0, [], 0)
        return combinations
    
    def combinationSum2(self, candidates: list[int], target: int) -> list[list[int]]:
        def backtrack(start, builder: list[int]):
            # found
            if sum(builder) == target:
                res.append(list(builder))

            # explore
            for i in range(start, len(candidates)):
                if sum(builder) + candidates[i] <= target:
                    builder.append(candidates[i])
                    backtrack(i, builder)
                    builder.pop()

        res = []
        backtrack(0, [])
        return res

File: leet/backtracking/combination_sum.cpp
// https://leetcode.com/problems/combination-sum

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int>* candPtr;
    vector<vector<int>> combinations;
    vector<int> builder;
    int n;
    int target;

    vector<vector<int>> combinationSum(vector<int>& candidates, int t) {
        candPtr = &candidates;
        n = candidates.size();
        target = t;

        sort(candidates.begin(), candidates.end());
        
        bt(0, 0);
        return combinations;
    }

    void bt(int start, int sum) {
        if (start == n)
            return;
        if (sum == target)
            combinations.push_back(builder);
        
        vector<int>& candidates = *candPtr;
        for (int i = start; i < n; i++) {
            int c = candidates[i];
            if (c+sum <= target) {
                sum += c;
                builder.push_back(c);
                bt(i, sum);
                sum -= c;
                builder.pop_back();
            }
        }
    }
};